### ISSUE TITLE
New manager: Bicep

### ISSUE THREAD

**Issue by felpel** [renovatebot/renovate#17484]
### What would you like Renovate to be able to do?

Bicep is a recently introduced IaC tool by Microsoft which allow developers to deploy their Azure resources ([docs](https://docs.microsoft.com/en-us/azure/azure-resource-manager/bicep/overview?tabs=bicep)).

It would be neat if Renovate could do the following:

## Resources & API version

Azure Resources rely on API versions for deployments. It is recommended by Microsoft to use the latest version of a given API, if applicable.

For instance, here is a declaration of a storage account with the `2021-04-01` API version:
```bicep
resource sa 'Microsoft.Storage/storageAccounts@2021-04-01' = {
  name: storageAccountName
  location: location
  sku: {
    name: 'Standard_LRS'
  }
  kind: 'StorageV2'
  properties: {
    accessTier: 'Hot'
  }
}
```

If Azure would introduce a new API version for a given resource type, it would be neat to have Renovate upgrade this specific API version.

## Modules & registries

Modules are re-usable blocks where users can create resources with default values and overwrite specific properties through configuration.

Modules can be defined locally in the same repository, but can also be found in a [remote registry](https://docs.microsoft.com/en-us/azure/azure-resource-manager/bicep/modules#file-in-registry).

For instance, here is a custom remote module to setup a storage account located in a ACR where `v1` is the set version:

```bicep
module stgModule 'br:exampleregistry.azurecr.io/bicep/modules/storage:v1' = {
  name: 'storageDeploy'
  params: {
    storagePrefix: 'examplestg1'
  }
}
```

If there would be a new version, Renovate could upgrade it in order to not lag behind.

### If you have any ideas on how this should be implemented, please tell us here.

N/A

### Is this a feature you are interested in implementing yourself?

No

**Comment by JamieMagee** [comment]
For resources and API versions the source of truth is this repository: https://github.com/Azure/azure-rest-api-specs/tree/main/specification

**Comment by felpel** [comment]
After an internal consultation with my peers who might have been interested with such feature, it didn't seem like Bicep would be a good candidate for Renovate after all.

Feel free to continue on this path @JamieMagee. If you would like, I have a draft for the questionnaire and I could email it to you.

**Comment by JamieMagee** [comment]
@felpel could you elaborate about why you don't think it would be a good candidate for Renovate?

**Comment by slapointe** [comment]
Azure Resource Providers' API versions are based on dates and not a versioning scheme like Semantic versioning where we can filter out breaking and non breaking changes.

A new API version from a RP is very likely to incorporate breaking changes. Auto updating these API in the context of infrastructure as code would be an anti-pattern.

Not speaking about all the "-preview" APIs that stays forever there and are kinda regular releases.

Azure need to clean and rework their versioning strategy before it could work as proposed by @felpel

**Comment by rarkins** [comment]
Automerging breaking updates is always an anti-pattern, but auto updating them is not. Renovate can raise PRs - or at least display them in a dependency dashboard if one exists - so that users are aware that new versions exist and they should consider migrating.

**Comment by rarkins** [comment]
@skyfrk the questionnaire linked above is an important start

**Comment by rarkins** [comment]
Looks good to start! Recommend starting with datasource, then versioning, then manager

**Comment by skyfrk** [comment]
@rarkins I built the GitHub GraphQL queries that I would like to use in the new data source in the mean time (see update questionnaire above).

I also saw that some data sources are based on weekly data dumps (https://github.com/renovatebot/renovate/tree/main/tools/static-data). 

Now I am wondering which way I should implement the new data source:

1. Adding another weekly data dump, which stores all provider namespaces and their available versions and just access the data dump in the new data source.
2. Calling the GitHub GraphQL endpoint and caching the results when looking up available versions in the new data source.

I guess the second option is preferred, because it provides a better user experience by offering more recent API versions?
